Support for `--add-host host.docker.internal:host-gateway`

Since Docker 20.10.0 it is now possible to specify `--add-host host.docker.internal:host-gateway` which will add an entry to the `/etc/hosts` file of the container like:

```
172.17.0.1      host.docker.internal
```

Which makes it easier for the Container to communicate with the Host.

I tried to use the `extraHosts` option of the docker-maven-plugin like this:

```
 <extraHosts>
    <host>host.docker.internal:host-gateway</host>
</extraHosts>
```

Unfortunately that yields the following error:

```
Caused by: java.lang.IllegalArgumentException: unable to resolve ip address for host-gateway
    at io.fabric8.maven.docker.access.ContainerHostConfig.extraHosts (ContainerHostConfig.java:97)
    at io.fabric8.maven.docker.service.RunService.createContainerHostConfig (RunService.java:420)
    at io.fabric8.maven.docker.service.RunService.createContainerConfig (RunService.java:379)
    at io.fabric8.maven.docker.service.RunService.createContainer (RunService.java:152)
    at io.fabric8.maven.docker.service.RunService.createAndStartContainer (RunService.java:190)
    at io.fabric8.maven.docker.service.helper.StartContainerExecutor.startContainer (StartContainerExecutor.java:47)
    at io.fabric8.maven.docker.StartMojo.lambda$startImage$0 (StartMojo.java:298)
    at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly (TrustedListenableFutureTask.java:125)
    at com.google.common.util.concurrent.InterruptibleTask.run (InterruptibleTask.java:69)
    at com.google.common.util.concurrent.TrustedListenableFutureTask.run (TrustedListenableFutureTask.java:78)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at com.google.common.util.concurrent.MoreExecutors$DirectExecutorService.execute (MoreExecutors.java:321)
    at java.util.concurrent.ExecutorCompletionService.submit (ExecutorCompletionService.java:181)
    at io.fabric8.maven.docker.StartMojo.startImage (StartMojo.java:296)
    at io.fabric8.maven.docker.StartMojo.executeInternal (StartMojo.java:165)
    at io.fabric8.maven.docker.AbstractDockerMojo.execute (AbstractDockerMojo.java:280)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    ...
Caused by: java.net.UnknownHostException: host-gateway: Name or service not known
    at java.net.Inet6AddressImpl.lookupAllHostAddr (Native Method)
    at java.net.InetAddress$2.lookupAllHostAddr (InetAddress.java:929)
    at java.net.InetAddress.getAddressesFromNameService (InetAddress.java:1324)
    at java.net.InetAddress.getAllByName0 (InetAddress.java:1277)
    at java.net.InetAddress.getAllByName (InetAddress.java:1193)
    at java.net.InetAddress.getAllByName (InetAddress.java:1127)
    at java.net.InetAddress.getByName (InetAddress.java:1077)
    at io.fabric8.maven.docker.access.ContainerHostConfig.extraHosts (ContainerHostConfig.java:95)
    at io.fabric8.maven.docker.service.RunService.createContainerHostConfig (RunService.java:420)
    at io.fabric8.maven.docker.service.RunService.createContainerConfig (RunService.java:379)
    at io.fabric8.maven.docker.service.RunService.createContainer (RunService.java:152)
    at io.fabric8.maven.docker.service.RunService.createAndStartContainer (RunService.java:190)
    at io.fabric8.maven.docker.service.helper.StartContainerExecutor.startContainer (StartContainerExecutor.java:47)
    at io.fabric8.maven.docker.StartMojo.lambda$startImage$0 (StartMojo.java:298)
    at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly (TrustedListenableFutureTask.java:125)
    at com.google.common.util.concurrent.InterruptibleTask.run (InterruptibleTask.java:69)
    at com.google.common.util.concurrent.TrustedListenableFutureTask.run (TrustedListenableFutureTask.java:78)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at com.google.common.util.concurrent.MoreExecutors$DirectExecutorService.execute (MoreExecutors.java:321)
    at java.util.concurrent.ExecutorCompletionService.submit (ExecutorCompletionService.java:181)
    at io.fabric8.maven.docker.StartMojo.startImage (StartMojo.java:296)
    at io.fabric8.maven.docker.StartMojo.executeInternal (StartMojo.java:165)
    at io.fabric8.maven.docker.AbstractDockerMojo.execute (AbstractDockerMojo.java:280)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    ...
```

### Info

* docker-maven-plugin version : 0.39.0
* Maven version (`mvn -v`) :
```
Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537)
Maven home: /home/aretter/.sdkman/candidates/maven/current
Java version: 1.8.0_312, vendor: BellSoft, runtime: /home/aretter/.sdkman/candidates/java/8.0.312.fx-librca/jre
Default locale: en_GB, platform encoding: UTF-8
OS name: "linux", version: "5.13.0-28-generic", arch: "amd64", family: "unix"
```
* Docker version : 20.10.12, build e91ed57

--- a/doc/changelog.md
+++ b/doc/changelog.md
@@ -1,5 +1,7 @@
 # ChangeLog
 * **0.44-SNAPSHOT**:
+  - Resolve registry auth URL by registry ID ([1688](https://github.com/fabric8io/docker-maven-plugin/issues/1688)) @wajda
+  - Allow 'host-gateway' as non-resolvable hostname in extraHosts ([1527](https://github.com/fabric8io/docker-maven-plugin/issues/1527)) @azaaiman
 
 * **0.43.0** (2023-05-20):
   - Make buildx build single non-native platform if requested ([1665](https://github.com/fabric8io/docker-maven-plugin/pull/1665)) @martyvona

--- a/src/main/java/io/fabric8/maven/docker/access/ContainerHostConfig.java
+++ b/src/main/java/io/fabric8/maven/docker/access/ContainerHostConfig.java
@@ -16,6 +16,7 @@ import io.fabric8.maven.docker.util.JsonFactory;
 
 public class ContainerHostConfig {
 
+    private static final String HOST_GATEWAY = "host-gateway";
     final JsonObject startConfig = new JsonObject();
 
     public ContainerHostConfig() {}
@@ -98,11 +99,7 @@ public class ContainerHostConfig {
                     throw new IllegalArgumentException("extraHosts must be in the form <host:host|ip>");
                 }
 
-                try {
-                    mapped.add(i, parts[0] + ":" + InetAddress.getByName(parts[1]).getHostAddress());
-                } catch (UnknownHostException e) {
-                    throw new IllegalArgumentException("unable to resolve ip address for " + parts[1], e);
-                }
+                mapped.add(i, parts[0] + ":" + resolveHostAddress(parts[1]));
             }
             return addAsArray("ExtraHosts", mapped);
         }
@@ -230,6 +227,17 @@ public class ContainerHostConfig {
             startConfig.add(propKey, JsonFactory.newJsonArray(props));
         }
         return this;
+    }
+
+    private String resolveHostAddress(String host) {
+        if (HOST_GATEWAY.equalsIgnoreCase(host)) {
+            return host;
+        }
+        try {
+            return InetAddress.getByName(host).getHostAddress();
+        } catch (UnknownHostException e) {
+            throw new IllegalArgumentException("unable to resolve ip address for " + host, e);
+        }
     }
 
     private ContainerHostConfig add(String name, String value) {

--- a/src/test/java/io/fabric8/maven/docker/access/ContainerHostConfigTest.java
+++ b/src/test/java/io/fabric8/maven/docker/access/ContainerHostConfigTest.java
@@ -41,6 +41,14 @@ class ContainerHostConfigTest {
     }
 
     @Test
+    void testExtraHostsAcceptsHostGateway() {
+        ContainerHostConfig hc = new ContainerHostConfig();
+        hc.extraHosts(Collections.singletonList("host.docker.internal:host-gateway"));
+
+        Assertions.assertEquals("{\"ExtraHosts\":[\"host.docker.internal:host-gateway\"]}", hc.toJson());
+    }
+
+    @Test
     void testUlimits() {
         Object data[] = {
             "{Ulimits: [{Name:bla, Hard:2048, Soft: 1024}]}", "bla", 2048, 1024,

